test/SILOptimizer/generalized_accessors.swift

```diff
@@ -0,0 +1,277 @@
+struct TestConvolutedPaths {
+  var flag: Bool
+  var stored : Int
+
+  var computed: Int {
+    _read {
+      if flag {
+        yield stored // expected-note {{previous yield was here}}
+      }
+      if !flag {
+        yield stored // expected-error {{encountered multiple yields along a single path in yield_once accessor}}
+      }
+    }
+
+    _modify {
+      if flag {
+        yield &stored // // expected-note {{previous yield was here}}
+      }
+      if !flag {
+        yield &stored // expected-error {{encountered multiple yields along a single path in yield_once accessor}}
+      }
+    }
+  }
+}
+
+struct TestYieldInLoops {
+  var stored: Int
+  var count: Int
+  var computed: Int {
+    _read {
+      for _ in 0..<count {
+        yield stored // expected-error {{encountered multiple yields along a single path in yield_once accessor}}
+                     // expected-note@-1 {{previous yield was here}}
+      }
+    }
+
+    _modify {
+      yield &stored // expected-note {{previous yield was here}}
+      for _ in 0..<count {
+        yield &stored // expected-error {{encountered multiple yields along a single path in yield_once accessor}}
+      }
+    }
+  }
+}
+
+enum SoftError : Error {
+  case Ignorable
+}
+func aThrowingFunction() throws {
+  throw SoftError.Ignorable
+}
+
+struct TestYieldInDoCatch {
+  var stored: Int
+  var computed: Int {
+    _read {
+      do {
+        try aThrowingFunction()
+        yield stored
+      } catch {
+        yield stored
+      }
+    }
+  }
+}
+
+struct TestYieldInDoCatch2 {
+  var stored: Int
+  var computed: Int {
+    _read {
+      do {
+        try aThrowingFunction()
+        yield stored
+      } catch {
+      }
+    } // expected-error {{cannot guarantee that the accessor yields a value before returning}}
+      // expected-note@-1 {{some paths reaching here have a yield and some don't}}
+
+    _modify {
+      do {
+        try aThrowingFunction()
+      }
+      catch {
+        yield &stored
+      }
+    } // expected-error {{cannot guarantee that the accessor yields a value before returning}}
+      // expected-note@-1 {{some paths reaching here have a yield and some don't}}
+  }
+}
+
+enum CompassPoint {
+  case north
+  case south
+  case east
+  case west
+}
+
+struct TestYieldInSwitch {
+  var stored: Int
+  var cp: CompassPoint
+
+  var computed: Int {
+    get { return stored }
+    _modify {
+      switch cp {
+      case .north:
+        yield &stored
+      case .south:
+        stored = 10
+      case .east:
+        yield &stored
+      case .west:
+        stored = 12
+      }
+      cp = .north // expected-note {{some paths reaching here have a yield and some don't}}
+    } // expected-error {{cannot guarantee that the accessor yields a value before returning}}
+  }
+}
+
+struct TestExplicitReturn {
+  var stored: Int
+  var flag: Bool
+
+  var computed: Int {
+    mutating _read {
+      if stored > 0 {
+        return
```

Comment:
Yes. In the following example, the diagnostic will go on the return statement

```_read {
      var loc = 1
       if flag {
         loc = 2
       }
       return // expected-error {{accessor doesn't yield before returning}}
    }
```
There can be only one return instruction in SIL. If the function can exit at multiple places the error will appear on the closing brace of the function definition. This is because at SIL level we only see one return instruction whose `Loc` is the closing brace of the function definition.